Improve COBS & add unit test cases for Backend/Frames

rohanpurohit opened the pull request:
This PR includes:
- Update COBS interface to fix resizing and change encode_cobs algorithm to account for edge cases for eg: a 254-byte buffer and buffers of size greater than 255.
- rename the read_ntoh function to read_bigend for better understanding.
- add Unit test cases for backend/Frames.h

rohanpurohit committed:
Driver: Add unit tests for Backend/Frames

rohanpurohit committed:
Apply clang-tidy fixes

rohanpurohit committed:
frames: add case for input_overwritten

rohanpurohit committed:
Frames: update after latest changes

rohanpurohit committed:
frames: consolidate test cases

rohanpurohit committed:
Merge remote-tracking branch 'origin/develop' into feature/firmware-test-driver-frames

rohanpurohit commented:
@ethanjli PR is ready for review. thanks

ethanjli commented on firmware/ventilator-controller-stm32/Core/Test/Pufferfish/Driver/Serial/Backend/frames.txt:
```diff
@@ -0,0 +1,138 @@
+   THEN("The output method reports ok status")
+   THEN("The output buffer has an expected sequence of 128 bytes")
+
+  WHEN("256-byte frame body with a delimiter at the 129th and 256th-byte position is passed as input, and output is called after that")
```
WHEN 256 bytes are passed as input, where the input's delimiter bytes are at positions 129 and 256, and then output is called afterwards

ethanjli reviewed (commented):
Tests to add:

- Correct resetting of FrameReceiver behavior after calling output from a valid input (like in chunks.txt)
- Correct resetting of FrameReceiver behavior after calling output from an input error (like in chunks.txt); cover input errors triggered by ChunkSplitter and errors triggered by COBSDecoder

rohanpurohit committed:
firmware: fix an edge case on a buffer size of 254 bytes with delimiter at the end

* for buffer sizes of length 254 bytes there are two possibilities:
1. when there are no delimiter bytes in the entire buffer or many null bytes except at the 254th byte postition
2. when there is a delimiter byte at the 254th byte position

in case 1: the encoded buffer will be of length 256 bytes
in case 2: the encoded buffer will be of length 255 bytes

catch this case in our cobs implementation.

rohanpurohit committed:
add code comments for non-returning error codes

rohanpurohit committed:
Add test cases for framereceiver after consuming output

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-test-driver-frames

rohanpurohit committed:
cobs: fix cobs encode implementation to handle 254 byte buffer edge case

rohanpurohit committed:
cobs: fix clang errors

rohanpurohit committed:
Endian: rename read_ntoh to read_bigend

rohanpurohit committed:
Cobs: fix encoded buffer resize logic and update test cases

rohanpurohit committed:
cobs: manually fix cppcheck warnings and clang errors

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/COBS.tpp:
```diff
@@ -39,25 +39,37 @@ IndexStatus encode_cobs(
-        break;
-      }
-      encoded_buffer_size = max_block_size + 2;
+  // calculate the exact encoded buffer size
```
As discussed, let's decompose the size-computation logic out into a separate function.

rohanpurohit committed:
cobs: modify the get_encoded_cobs_buffer_size method to calculate the actual encoded size

rohanpurohit committed:
update get_encoded_cobs_buffer_size function

rohanpurohit committed:
cobs: fix get_encoded_cobs_buffer_size function

- return IndexStatus instead of size

ethanjli commented on firmware/ventilator-controller-stm32/Core/Inc/Pufferfish/Util/COBS.tpp:
```diff
@@ -115,7 +116,7 @@ constexpr size_t get_encoded_cobs_buffer_size(
       write_index++;
     }
     if (read_index > unencoded_buffer.size()) {
```
What should happen if `read_index == unencoded_buffer.size()`? Right now it looks like then line 121 would try to do a memory access at `unencoded_buffer[unencoded_buffer.size()]`. Is that within the bounds of `unencoded_buffer`? (I think the answer is no, but I always get confused about potential off-by-one cases)

(also, would it be possible to generate a test case where this happens?)

rohanpurohit replied:
I think read_index is always less than `unencoded_buffer.size()` as it won't enter the while loop if it's equal to it. So in a way we are implicitly doing the bounds check, then the question is do we need to explicitly catch `out_of_bounds` ?

ethanjli replied:
Oops!!! I had been thinking of this function as reading encoded data, when clearly it's reading pre-encoded data. The only way we read data is checking if a byte is zero, and deciding based on that whether to increment some (index) counters. So the bounds check of the while loop is sufficient; invalid data has no way to force an out-of-bounds read.

rohanpurohit committed:
fix the get_encoded_buffer_size function implementation

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-test-driver-frames

rohanpurohit committed:
fix clang-tidy warnings

rohanpurohit commented:
1. This project is licensed under Apache License v2.0 for any software, and Solderpad Hardware License v2.1 for any hardware - do you agree that your contributions to this project will be under these licenses, too? Yes.
2. Were any of these contributions also part of work you did for an employer or a client? No.
3. Does this work include, or is it based on, any third-party work which you did not create? No.

ethanjli commented:
Let's merge the latest version of develop back into this branch, and then try running the firmware with these changes to ensure that there are no regressions or subtle merge conflicts. Afterwards, I'll mark the formal approval on Github.

rohanpurohit committed:
Merge branch 'develop' into feature/firmware-test-driver-frames

ethanjli reviewed (approved):
From light testing, it seems to work fine without any issues.
